rfcs/0007-pass-spec.md

```diff
@@ -0,0 +1,453 @@
+In particular, the restrictions about adding new widgets may be too constraining.
+
+However, I'd argue this is a rigidity that already existed, but simply was undocumented:
+
+- Druid and Masonry have always worked on a pass system, with the passes being based on recursing the Widget methods.
+- Druid's passes were already constraining, with the Notification and Command events being included as escape hatches to escape some of these constraints.
+- Adding a new widget in the middle of a layout pass and trying to lay out that widget could already crash the application or trigger a warning.
+
+With this RFC's design the restrictions are documented and, wherever possible, enforced by the type system.
+
+
+## Rationale and alternatives
+
+The main benefit of this RFC is to untangle code.
+
+By specifying what passes do and how they interact with each other, we can make pass-related code more structured and modular.
+
+We could make a less ambitious RFC where we document the widget behavior more formally, but avoid some major changes:
+
+- Keep the current "`MyParentWidget::on_text_event` must call `WidgetPod<MyChildWidget>::on_text_event` for every child" approch.
+- Cut the part where mutating the widget tree is only allowed in `MutateCtx`.
+
+However, these major changes both help us massively simplify Masonry's internal code:
+
+- Having `RenderRoot` be in charge of event targeting means we can get rid of the WidgetPod's "should this event be recursed to the inner widget" logic.
+- Having direct event targeting means we can remove a bunch of `RouteFoobar` events that do nothing but carry a `Foobar` event through the widget tree.
+- Limiting mutations to the widget tree to a specific pass lets us remove the `WidgetAdded` event and some very thorny logic.
+
+Overall, these changes together will make the new pass system more cohesive and predictable.
+
+
+## Prior art
+
+TODO
+
+- Bevy spawn
+- Compose pass: web platform.
+- pass systems in other Rust frameworks
+- Qt
+
+
+## Unresolved questions
+
+The RFC is vague on the following implementation points:
+
+- Exact pointer capture behavior.
+- Exact focus behavior.
+- Detailed description of the UPDATE passes.
+- Static limit on the number of reruns.
+- The total list of context methods.
+
+We will likely figure them out during implementation and document them in real-time.
+
+
+## Future possibilities
+
+### Reparenting children
+
+Right now children can only be added and removed.
+
+In the future, we could add the ability to splice a widget subtree from one parent widget to another.
+
+### Focus
+
+Focus needs to be better defined.
+
+The concept of "local focus" vs "active focus" might need a more formal definition.
+
+A related question is how we preserve local focus for situations that "borrow" it. For instance, if you press `Tab` to open a menu, the menu should have focus.
```

Comment:
When does tab open a menu? Do you mean <kbd>alt</kbd> here?

(Incidentally, GitHub supports the `<kbd>` tag)

Reply:
Oops, yeah, I meant Alt.